pyuvdata/tests/__init__.py

```diff
@@ -5,19 +5,245 @@
+                            )
+                        )
+                    elif self.match is not None:
+                        for record in self:
+                            if str(record.message).startswith(
+                                self.match[warn_i]
+                            ) or re.compile(self.match[warn_i]).search(
+                                str(record.message)
+                            ):
+                                if issubclass(
+                                    record.category, self.expected_warning[warn_i]
+                                ):
+                                    break
+                        else:
+                            raise AssertionError(
+                                f"No warnings of type {self.expected_warning[warn_i]} "
+                                f"matching ('{self.match[warn_i]}') was "
+                                "emitted. The list of emitted warnings is: "
+                                f"{[each.message for each in self]}."
+                            )
+
+
+def check_warnings(expected_warning, match, nwarnings=None, *args, **kwargs):
+    """
+    Assert that code raises a particular set of warnings, used as a context manager.
+
+    Similar to ``pytest.warns``, but allows for specifying multiple warnings.
+    It also better matches warning strings when the warning uses f-strings or
+    formating.
+
+    Note that unlike the older checkWarnings function, the warnings can be passed
+    in any order, they do not have to match the order the warnings are raised
+    in the code.
+
+    Parameters
+    ----------
+    expected_warning : list of Warnings or Warning
+        List of expected warnings. If a single warning type or a length 1 list,
+        will be used for the type of all warnings.
+    match : list of str or str
+        List of strings to match warnings to. If a str or a length 1 list, will
+        be used for all warnings.
+    nwarnings : int, optional
+        Option to specify that multiple of a single type of warning is expected.
+        Only used if category and match both only have one element.
+
+    """
+    __tracebackhide__ = True
+    if not isinstance(expected_warning, list):
+        expected_warning_list = [expected_warning]
+    else:
+        expected_warning_list = expected_warning
+    if not isinstance(match, list):
+        match_list = [match]
+    else:
+        match_list = match
+
+    if (
+        len(expected_warning_list) > 1
+        and len(match_list) > 1
+        and len(expected_warning_list) != len(match_list)
+    ):
+        raise ValueError(
+            "If expected_warning and match both have more than one element, "
+            "they must be the same length."
+        )
+
+    if len(expected_warning_list) > 1 or len(match_list) > 1:
+        nwarnings = max(len(expected_warning_list), len(match_list))
+    elif nwarnings is None:
+        nwarnings = 1
```

Comment:
This block is the first time `nwarnings` appears, and it overwrites the definition of the variable if there is more than one warning. Should we even have it as an optional kwarg if we discard user input?

Reply:
So the point (in my head) of `nwarnings` is the case where you have a lot of only one warning over and over. In the docstring I tried to make it clear that `nwarnings` was only used if both the error and message only had one element. I don't like that with the current function you have to specify nwarnings when you've already specified how many by listing all the errors and messages.

Reply:
Ah okay I think I get it now. I was not necessarily thinking about this case when I read the block.